### PR TITLE
Run journal in containers

### PR DESCRIPTION
**Pull request by giorgiosironi**
Main part of https://github.com/elifesciences/issues/issues/3676

A host-based nginx is the only remaining part since it has a complex configuration. It forwards to static files in the `elifesciences/journal_web` container, and to the php-fpm container `elifesciences/journal`.

### CODE REVIEW

**Comment by giorgiosironi** [comment]
Build still missing `elifesciences/journal_web:latest`, as expected, but no other errors up to that point.

File: salt/example.top [hunk 1]
```diff
@@ -1,14 +1,7 @@
...
+        - elife.docker
         - elife.nginx
-        - elife.nginx-php7
         - elife.nginx-error-pages
-        - elife.nodejs6
-        - elife.redis-server
-        - journal.curl-7-36
         - journal
-        - api-dummy
-        - journal.local-demo
```
Comment (giorgiosironi) [inline salt/example.top]: lots of :wastebasket:

File: salt/journal/config/srv-journal-containers.env [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{%- if pillar.elife.env == 'dev' -%}
+{%- set app_env = 'ci' -%}
+{%- else -%}
+{%- set app_env = pillar.elife.env -%}
+{%- endif -%}
+APP_ENV={{ app_env }}
```
Comment (giorgiosironi) [inline salt/journal/config/srv-journal-containers.env]: is this actually useful? Wasn't provided before

File: salt/journal/config/srv-journal-nginx-assets.conf [hunk 3]
```diff
@@ -0,0 +1,9 @@
+server {
+    server_name localhost;
+    listen 8080;
+    root /srv/journal/web;
+
+    location / {
```
Comment (giorgiosironi) [inline salt/journal/config/srv-journal-nginx-assets.conf]: weird emptyness, but it works in serving the contents of `/srv/journal/web`.

**Comment by giorgiosironi** [comment]
`fresh` and its smoke tests are green, promising. Creating an instance for exploratory testing.

**Comment by giorgiosironi** [comment]
https://pr-90-manual--journal.elifesciences.org/ is built with the latest version of this branch.

File: salt/journal/config/srv-journal-containers.env (hunk 2 above)
Comment (giorgiosironi) [reply salt/journal/config/srv-journal-containers.env]: Wrong, it was provided before through `fastcgi_param` and in various other places.

**Comment by lsh-0** [comment]
(ignore me, just including myself in updates)

**Comment by lsh-0** [comment]
this: https://github.com/elifesciences/journal-formula/pull/90/files#diff-f165c8e699a3e12e5e7454d26ee174cdL209

also exists here:
https://github.com/elifesciences/builder-base-formula/blob/master/elife/nginx.sls#L105-L113

the journal-formula is using a native salt state, so it should probably replace the one in nginx.sls

**Comment by lsh-0** [comment]
the `--no-ansi` flags I added prevents some of the noise in the highstate

**Comment by lsh-0** [comment]
the containerisation branch builds fine in ubuntu 18.04 after a little debugging - I was using an unclean vagrant machine that still had that local-demo.conf nginx conf hanging around interfering with port 8080, and I got some errors from GPG keys and the Docker ppa, so I opened a PR for builder-base to fix those.

using `/etc/init.d/nginx` to do the reload and put it into 'maintenance mode' is kinda flaky, but I'm more concerned about getting it deployed on 18.04 as quickly and as safely as possible. Pushing a chunk of the machinery into the container will help this.

**Comment by lsh-0** [comment]
this was the only error on a clean 18.04 vagrant build of this branch:

```
    journal--vagrant: ----------
    journal--vagrant:           ID: journal-cache-warmup
    journal--vagrant:     Function: cmd.run
    journal--vagrant:         Name: docker-compose --no-ansi exec fpm bin/console cache:warmup
    journal--vagrant:       Result: False
    journal--vagrant:      Comment: Command "docker-compose --no-ansi exec fpm bin/console cache:warmup" run
    journal--vagrant:      Started: 07:25:49.774255
    journal--vagrant:     Duration: 614.658 ms
    journal--vagrant:      Changes:   
    journal--vagrant:               ----------
    journal--vagrant:               pid:
    journal--vagrant:                   27666
    journal--vagrant:               retcode:
    journal--vagrant:                   1
    journal--vagrant:               stderr:
    journal--vagrant:                   the input device is not a TTY
    journal--vagrant:               stdout:
    journal--vagrant: ----------
```

can't replicate it with subsequent highstate. Going to remove that `--no-ansi` from the command though

**Comment by giorgiosironi** [comment]
> can't replicate it with subsequent highstate. Going to remove that --no-ansi from the command though

`docker exec -t` may (or may not) help.

**Comment by giorgiosironi** [comment]
Note that my goal is to merge this on the existing 14.04 servers, as 18.04 servers can then replace them easily.

**Comment by giorgiosironi** [comment]
> using /etc/init.d/nginx to do the reload and put it into 'maintenance mode' is kinda flaky

The ELB detaches the servers when `nginx` is stopped, for zero downtime

> update: oooh, it's pillar.journal.web_users vs pillar.elife.web_users. definitely a wart that needs lancing

They're not identical though. Ideal for after the migration.

**Comment by giorgiosironi** [comment]
Forgot smoke tests, which are removed along with the repository. Adding to the formula.

**Comment by giorgiosironi** [comment]
Smoke tests pass locally but will need some more debugging work here. These tests run with the `continuumtest` environment.

File: salt/example.top (hunk 1 above)
Comment (thewilkybarkid) [reply salt/example.top]: 👋

File: salt/journal/config/srv-journal-docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            - "9000:9000"
+        env_file:
+            - ./containers.env
+        restart: always
+        volumes:
+            - ./parameters.yml:/srv/journal/app/config/parameters.yml
+        depends_on:
+            - redis
+    assets:
+        image: "elifesciences/journal_web:${IMAGE_TAG}"
```
Comment (thewilkybarkid) [inline salt/journal/config/srv-journal-docker-compose.yml]: Can we avoid having two published images? Refs https://github.com/elifesciences/journal/pull/1032.

File: salt/journal/init.sls [hunk 5]
```diff
@@ -6,48 +6,61 @@ maintenance-mode-start:
...
+        - group: {{ pillar.elife.deploy_user.username }}
+        - template: jinja
+        - require:
+            - journal-folder
 
 config-file:
     file.managed:
-        - name: /srv/journal/app/config/parameters.yml
+        - name: /srv/journal/parameters.yml
+        # TODO: rename
```
Comment (thewilkybarkid) [inline salt/journal/init.sls]: The source file?

File: salt/journal/config/srv-journal-docker-compose.yml (hunk 4 above)
Comment (giorgiosironi) [reply salt/journal/config/srv-journal-docker-compose.yml]: I couldn't find the resolution of the original issues, so with this new information, maybe

File: salt/journal/init.sls (hunk 5 above)
Comment (giorgiosironi) [reply salt/journal/init.sls]: `salt://journal/config/srv-journal-app-config-parameters.yml` as it doesn't reflect the host's path anymore

**Comment by lsh-0** [comment]
> Note that my goal is to merge this on the existing 14.04 servers, as 18.04 servers can then replace them easily.

of course, it would be so easy otherwise

> The ELB detaches the servers when nginx is stopped, for zero downtime

ah, I didn't realise. Good to know.

File: salt/journal/config/etc-nginx-sites-available-journal.conf [hunk 6]
```diff
@@ -132,8 +119,17 @@ server {
         {% endif %}
     }
 
+    location = /favicon.ico {
+        proxy_pass http://localhost:8080/favicon.ico;
+    }
+
     location /assets {
         add_header Cache-Control "public, max-age={{ 60 * 60 * 24 * 365 }}, immutable";
```
Comment (thewilkybarkid) [inline salt/journal/config/etc-nginx-sites-available-journal.conf]: Should probably be in the `assets` Nginx (and combine the `location`s)?

File: salt/journal/config/srv-journal-docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+version: '3'
+
+services:
+    redis:
```
Comment (thewilkybarkid) [inline salt/journal/config/srv-journal-docker-compose.yml]: Won't be needed in `prod`?

Comment (giorgiosironi) [reply salt/journal/config/srv-journal-docker-compose.yml]: The formula however runs a build that won't create an ElastiCache (it takes longer). I think we can swap it in and out depending on whether ElastiCache is present.

File: salt/journal/config/etc-nginx-sites-available-journal.conf (hunk 6 above)
Comment (thewilkybarkid) [reply salt/journal/config/etc-nginx-sites-available-journal.conf]: In fact, do we even need a container for these? Could copy the first to the host instead?

Comment (giorgiosironi) [reply salt/journal/config/etc-nginx-sites-available-journal.conf]: Forward looking, containers are all there is as in an orchestrator there is no accessible host; trying to minimize what the host nginx does as eventually it should disappear.

Comment (thewilkybarkid) [reply salt/journal/config/etc-nginx-sites-available-journal.conf]: Fair enough. 👍

**Comment by thewilkybarkid** [comment]
We will need to monitor closely the performance, since the PHP-FPM configuration might need improving.

**Comment by giorgiosironi** [comment]
Load testing to be performed on a masterless server before merging.

**Comment by giorgiosironi** [comment]
Load testing the homepage with at a low level of concurrency, results in the same ballpark for a production server and a masterless stack with these changes:
```
Percentage of the requests served within a certain time (ms)
  50%    150
  66%    159
  75%    163
  80%    167
  90%    177
  95%    196
  98%    214
  99%    216
 100%    216 (longest request)
...
Percentage of the requests served within a certain time (ms)
  50%     97
  66%     99
  75%    100
  80%    101
  90%    104
  95%    106
  98%    172
  99%    191
 100%    191 (longest request)
```
Now turning up concurrency and sample size.

**Comment by giorgiosironi** [comment]
Same at c=4 n=1000, with `prod` being slowest as it is already handling real traffic.
```
Requests per second:    13.35 [#/sec] (mean)
Percentage of the requests served within a certain time (ms)
  50%    291
  66%    305
  75%    316
  80%    324
  90%    341
  95%    355
  98%    375
  99%    388
...
Requests per second:    19.67 [#/sec] (mean)
Percentage of the requests served within a certain time (ms)
  50%    198
  66%    200
  75%    202
  80%    204
  90%    209
  95%    242
  98%    271
  99%    284
```

**Comment by giorgiosironi** [comment]
By the way, cannot merge as https://github.com/elifesciences/issues/issues/4901#issuecomment-498126157 has broken `end2end`.

**Comment by giorgiosironi** [comment]
Ready to merge tomorrow morning, starting to apply it from `end2end` onwards.

**Comment by lsh-0** [comment]
end2end is green now for me, kicking this build

**Comment by lsh-0** [comment]
> Ready to merge tomorrow morning, starting to apply it from end2end onwards.

Ticket for that here if you're going to do the upgrade yourself: https://github.com/elifesciences/issues/issues/4918

**Comment by lsh-0** [comment]
the errors now look like our old friend php7.0 on 14.04

**Comment by giorgiosironi** [comment]
> Ticket for that here if you're going to do the upgrade yourself: elifesciences/issues#4918

I'm going to do the containers formula change for today, but I can also do the OS upgrade later.

> the errors now look like our old friend php7.0 on 14.04

It is; https://github.com/elifesciences/heavybox-formula/pull/10 fixed `builder-base-formula` builds not to use it, but we still want actual project formulas to fail if they are on "php7.0 on 14.04" as they can't work at all.

**Comment by giorgiosironi** [comment]
Temporarily disabling branch protection to allow the merge which will repair the build.

**Comment by lsh-0** [comment]
> but we still want actual project formulas to fail if they are on "php7.0 on 14.04" as they can't work at all.

of course.

I'm happy to do the os upgrade for journal, it just read like you were going to do it. It's multi-node so should be relatively safe, unlike api-gateway that I was hoping to do today until iiif stole all my time.

**Comment by giorgiosironi** [comment]
Also have to verify this didn't cause https://github.com/elifesciences/elife-xpub/issues/1984

**Pull request merged**